apps/arweave/src/ar_mining_worker.erl

```diff
@@ -245,281 +277,181 @@ terminate(_Reason, _State) ->
-						{partition_number2, Candidate2#mining_candidate.partition_number2},
-						{cm_peer, ar_util:format_peer(Candidate2#mining_candidate.cm_lead_peer)},
-						{cache_ref, Candidate2#mining_candidate.cache_ref},
-						{nonce, Candidate2#mining_candidate.nonce},
-						{session, ar_nonce_limiter:encode_session_key(SessionKey)}])
-			end,
-			State2
-	end.
+%% @doc Handle the `compute_h0` task.
+%% Indicates that the VDF step has been computed.
+handle_task({compute_h0, Candidate, _ExtraArgs}, State) ->
+	#state{
+		latest_vdf_step_number = LatestVDFStepNumber,
+		vdf_queue_limit = VDFQueueLimit
+	} = State,
+	#mining_candidate{ step_number = StepNumber } = Candidate,
+	State2 = report_and_reset_hashes(State),
+	% Check if we need to compute h0 early
+	case StepNumber >= LatestVDFStepNumber - VDFQueueLimit of
+		true ->
+			ar_mining_hash:compute_h0(self(), Candidate),
+			{noreply, State2#state{ latest_vdf_step_number = max(StepNumber, LatestVDFStepNumber) }};
+		false ->
+			{noreply, State2}
+	end;
 
-priority(computed_h2, StepNumber) ->
-	{1, -StepNumber};
-priority(computed_h1, StepNumber) ->
-	{2, -StepNumber};
-priority(compute_h2_for_peer, StepNumber) ->
-	{2, -StepNumber};
-priority(chunk2, StepNumber) ->
-	{3, -StepNumber};
-priority(chunk1, StepNumber) ->
-	{4, -StepNumber};
-priority(computed_h0, StepNumber) ->
-	{5, -StepNumber};
-priority(compute_h0, StepNumber) ->
-	{6, -StepNumber}.
+%% @doc Handle the `computed_h0` task.
+%% Indicates that the hash for the VDF step has been computed.
+handle_task({computed_h0, Candidate, _ExtraArgs}, State) ->
+	#mining_candidate{
+		h0 = H0, partition_number = Partition1, partition_upper_bound = PartitionUpperBound
+	} = Candidate,
+	{RecallRange1Start, RecallRange2Start} = ar_block:get_recall_range(H0, Partition1, PartitionUpperBound),
+	Partition2 = ar_node:get_partition_number(RecallRange2Start),
+	Candidate2 = generate_cache_ref(Candidate#mining_candidate{ partition_number2 = Partition2 }),
+	%% Check if the recall ranges are readable to avoid reserving cache space for non-existent data.
+	Range1Exists = ar_mining_io:is_recall_range_readable(Candidate2, RecallRange1Start),
+	Range2Exists = ar_mining_io:is_recall_range_readable(Candidate2, RecallRange2Start),
+
+	case {Range1Exists, Range2Exists} of
+		{true, true} ->
+			%% Both recall ranges are readable, so we need to reserve cache space for both.
+			case try_to_reserve_cache_space(2, Candidate2#mining_candidate.session_key, State) of
+				{true, State1} ->
+					%% Read the recall ranges; the result of the read will be reported by the `chunk1` and `chunk2` tasks.
+					ar_mining_io:read_recall_range(chunk1, self(), Candidate2, RecallRange1Start),
+					ar_mining_io:read_recall_range(chunk2, self(), Candidate2, RecallRange2Start),
+					{noreply, State1};
+				false ->
+					%% We don't have enough cache space to read the recall ranges, so we'll try again later.
+					add_delayed_task(self(), computed_h0, Candidate),
```

Comment:
See above comment. By the time we retry this task, the vdf step may be too old in which case we should abandon it so that the miner can focus on newer VDF steps (the older a VDF step is the greater the likelihood that any solution derived from it will be orphaned/invalid due to another miner mining a block with a higher step number)

Reply:
Oh, perhaps this comment from a while back got lost. This is the point I'm referencing [here](https://github.com/ArweaveTeam/arweave/pull/697/files#r1999936725)